### PR TITLE
Floorplanning Rework

### LINKED ISSUE #335: `FP_DEF_TEMPLATE` incomplete application
### Description

When setting `FP_DEF_TEMPLATE` the flow copies the diearea and io pins from the specified file. However this is not enough. If the design doesn't set `DIEA_AREA` to match the template def, the generated rows would be for the old design area before applying the template and this is certainly what's not expected. See the image below.
image

  

### Expected Behavior

The flow should properly handle applying a def template without needed additional variables.

### Environment report

```sh
n/a
```

### Reproduction material

n/a

### Relevant log output

```sh
n/a
```

### LINKED ISSUE #337: Remove `FP_SIZING`
### Description

The flow can easily detect whether a `DIEA_AREA` is specified or not and act accordingly making `FP_SIZING` redundant.   

### Proposal

Remove `FP_SIZING` and do the floorplan based on `DIE_AREA` and `CORE_AREA`

### LINKED ISSUE #338: `PL_TARGET_DENSITY_PCT` proper calculation
### Description

When `FP_SIZING` is `absolute`, `FP_CORE_UTIL` becomes irrelevant which is used by the flow to calculate `PL_TARGET_DENSITY_PCT`. In that case, The flow should calculate it differently.

### Proposal

Perhaps rely on metrics?

### PR DESCRIPTION
**Pull request by kareefardi**
* `Classic`
  * Removed `Odb.ApplyDEFTemplate`- now redundant to `Floorplan`
* `OpenROAD.Floorplan`
  * Cast obstruction to `int` after conversion to dbu. Fixing crash reported by @mo-hosni.
  * `FP_DEF_TEMPLATE` now handled by this step using OpenROAD's `read_def -floorplan_initialize` feature
  * `PL_TARGET_DENSITY_PCT` default calculation now prioritizes using metric `design__instance__utilization` if available.
  * Removed `FP_SIZING`
* `Odb.ApplyDEFTemplate`:
  * Deprecated
* `OpenROAD.GlobalPlacementSkipIO`:
  * Updated to do nothing when `FP_DEF_TEMPLATE` is defined
* `OpenROAD.IOPlacement`:
  * Updated to do nothing when `FP_DEF_TEMPLATE` is defined.
* `openlane.common.metrics`:
  * Add `*wns*` and `*tns*` to default comparison filter list.

## Testing
* Fix configurations that set `FP_SIZING` to `relative` and set a `DIE_AREA` 
* Explicitly set `PL_TARGET_DENSITY_PCT` to the value based of old calculation method as some designs had negative metric changes.

---

Depends on:
* https://github.com/efabless/openlane2-step-unit-tests/pull/22/
* https://github.com/efabless/openlane2-ci-designs/pull/17
---

Fixes #338
Fixes #337 
Fixes #335

### CODE REVIEW

**Comment by donn** [comment]
My point about "critical" creating noise for changes that are good is in full force here lol

**Review by donn — APPROVED** [review]

File: openlane/steps/openroad.py [hunk 1]
```diff
@@ -956,9 +986,12 @@ def get_script_path(self):
     def run(self, state_in: State, **kwargs) -> Tuple[ViewsUpdate, MetricsUpdate]:
         kwargs, env = self.extract_env(kwargs)
         if self.config["PL_TARGET_DENSITY_PCT"] is None:
-            expr = (
-                self.config["FP_CORE_UTIL"] + (5 * self.config["GPL_CELL_PADDING"]) + 10
-            )
+            util = self.config["FP_CORE_UTIL"]
+            metrics_util = state_in.metrics.get("design__instance__utilization")
+            if metrics_util:
```
Comment (donn) [inline openlane/steps/openroad.py:991]: is not None

File: openlane/steps/odb.py [hunk 2]
```diff
@@ -91,7 +91,7 @@ def get_command(self) -> List[str]:
                 "Misconfigured SCL: 'TECH_LEFS' must return exactly one Tech LEF for its default timing corner."
             )
 
-        lefs = ["--input-lef", tech_lefs[0]]
+        lefs: List[str] = ["--input-lef", str(tech_lefs[0])]
```
Comment (donn) [inline openlane/steps/odb.py:94]: Not 100% necessary since it can be inferred

File: openlane/steps/odb.py [hunk 3]
```diff
@@ -160,6 +160,9 @@ def get_command(self) -> List[str]:
         ]
 
     def run(self, state_in, **kwargs) -> Tuple[ViewsUpdate, MetricsUpdate]:
+        warn(
+            f"{self.id} is now deprecated. Template def is now applied in OpenROAD.Floorplan."
```
Comment (donn) [inline openlane/steps/odb.py:164]: Why not `@deprecated`?

File: openlane/steps/openroad.py [hunk 4]
```diff
@@ -715,9 +720,28 @@ class Floorplan(OpenROADStep):
         ),
     ]
 
+    class Mode(str, Enum):
```
Comment (donn) [inline openlane/steps/openroad.py:723]: Feel like an enum is overkill here instead of `Literal['template', 'absolute', 'relative']`

**Review by donn — CHANGES_REQUESTED** [review]
disregard the previous approval

File: openlane/steps/openroad.py (hunk 4 above)
Comment (kareefardi) [reply openlane/steps/openroad.py:723]: I think it is better if we decide to change the internal name (value of the properties) later.

**Review by donn — APPROVED** [review]

**Pull request merged**